src/matrix/kaldi-matrix.cc

```diff
@@ -2784,6 +2784,78 @@ void MatrixBase<Real>::DiffTanh(const MatrixBase<Real> &value,
 
 
 template<typename Real>
+void MatrixBase<Real>::AddSpatialRegularizationDeriv(
+    const MatrixBase<Real>& out_value, Real scale, Real* regularization_sumsq) {
+  const unsigned int kBlockSize = 256;
```

Comment:
please move this function to the 'else' in the cu-matrix function, or to a static non-member function defined in cu-matrix.cc; I don't like to bulk up the kaldi-matrix library with these neural-net-specific things.  At some point these may get moved out of the cu-matrix class.